test/programs/RelationalPrograms.jl

```diff
@@ -66,6 +69,47 @@ set_junction!(d, [1,4,2,4,3,4])
 set_junction!(d, [1,2,3], outer=true)
 @test parsed == d
 
+
+# Typed by Ints
+#------
+
+parsed = @relation (x,y,z) where (x::1, y::2, z::3, w::4) begin
+  R(x,w)
+  S(y,w)
+  T(z,w)
+end
+
+d = RelationDiagram([:1,:2,:3])
+add_box!(d, [:1,:4], name=:R)
+add_box!(d, [:2,:4], name=:S)
+add_box!(d, [:3,:4], name=:T)
+add_junctions!(d, [:1,:2,:3,:4], variable=[:x,:y,:z,:w])
+set_junction!(d, [1,4,2,4,3,4])
+set_junction!(d, [1,2,3], outer=true)
+@test parsed == d
+
+
+
+# Typed by Expressions
+#------
+
+parsed = @relation (x,y,z) where (x::n(1), y::n(2), z::n(3), w::n(4)) begin
+  R(x,w)
+  S(y,w)
+  T(z,w)
+end
+
+# Testing doesn't work right now because Julia doesn't like calling n as a function
+# d = RelationDiagram([Expr(n(1)), Expr(n(1)), Expr(n(1))])
+# add_box!(d, [:n(1),:n(4)], name=:R)
+# add_box!(d, [:n(2),:n(4)], name=:S)
+# add_box!(d, [:n(3),:n(4)], name=:T)
+# add_junctions!(d, [:n(1),:n(2),:n(3),:n(4)], variable=[:x,:y,:z,:w])
+# set_junction!(d, [1,4,2,4,3,4])
+# set_junction!(d, [1,2,3], outer=true)
+# @test parsed == d
+
```

Comment:
Mixed Symbol and expression case.

```julia
parsed = @relation (x,y,z) where (x::X, y::n(2), z::n(3)+X, w::n(4)) begin
  R(x,w)
  S(y,w)
  T(z,w)
end
```
This will probably error because of the overly strict typing above.

Mixed symbol, number, expression case:


```julia
parsed = @relation (x,y,z) where (x::X, y::n(2), z::n(3)+X, w::4) begin
  R(x,w)
  S(y,w)
  T(z,w)
end
```

This will definitely error because of the way you are checking element types. Basically the expression case should be handling any subtype of `Union{Symbol, Int, Expr}` as the element type. And then just think of :foo or :(4) as the expression containing a symbol or number.